### PR TITLE
Improved docs

### PR DESCRIPTION
**Pull request by fjetter**
# Description:

This reworks our documentation and introduces an API page where the most commonly used functions are referenced. This should allow users much more easily to get started and find things

### CODE REVIEW

**Comment by fjetter** [comment]
Builds fail because I already included a section about #226 which is not, yet, part of this PR

File: docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    read_dataset_as_metapartitions__iterator
+
+
+Dask
+****
+
+The `dask` module offers a seamless integration to `dask <https://docs.dask.org>`_
+and offers implementations for dask data collections like `dask.Bag`,
+`dask.DataFrame` or as `dask.Delayed`.
+This implemenation is best suited to handle big data and scale the
```
Comment (lr4d) [inline docs/api.rst]:
```suggestion
This implementation is best suited to handle big data and scale the
```

File: docs/guide/mutating_datasets.rst [hunk 2]
```diff
@@ -1,64 +1,19 @@
...
-more efficient as only a select amount of files need to be read.
+It's possible to update existing data
+by adding new physical partitions to them and deleting or replacing old partitions. Kartothek
+provides update functions that generally have the prefix `update_dataset` in their names.
+For example, :func:`~kartothek.io.eager.update_dataset_from_dataframes` is the update
+function for the ``eager`` backend.
 
-To see explicit partitioning in action, let's set up some data and a storage location
-first and store the data there with Kartothek:
+To see updating in action, let's first set up a storage location first and store
```
Comment (lr4d) [inline docs/guide/mutating_datasets.rst]:
```suggestion
To see updating in action, let's first set up a storage location and store
```

File: docs/guide/mutating_datasets.rst [hunk 3]
```diff
@@ -436,7 +310,7 @@ periodically to decrease unnecessary storage use.
 
 An example of garbage collection is shown below.
 A little above, near the end of the delete section,
-we removed two partitions for the dataset with uuid `another_partitioned_dataset`.
+we removed two partitions for the dataset with uuid `partitioned_dataset`.
```
Comment (lr4d) [inline docs/guide/mutating_datasets.rst]: `dm.uuid == "replace_partition"` oder?

File: docs/guide/mutating_datasets.rst [hunk 4]
```diff
@@ -448,7 +322,7 @@ When garbage collection is called, the files are removed.
 
     files_before = set(store.keys())
 
-    garbage_collect_dataset(store=store_factory, dataset_uuid="another_partitioned_dataset")
+    garbage_collect_dataset(store=store_factory, dataset_uuid=dm.uuid)
```
Comment (lr4d) [inline docs/guide/mutating_datasets.rst]: here, did you verify the set difference output below?

File: docs/guide/partitioning.rst [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    dm = update_dataset_from_ddf(
+        ddf, dataset_uuid="no_shuffle", store=store_factory, partition_on="A", table="table"
+    ).compute()
+    sorted(dm.partitions.keys())
+
+Shuffling
+*********
+
+To circumvent this, we offer a shuffle implementation for dask dataframes.
```
Comment (lr4d) [inline docs/guide/partitioning.rst]: > To circumvent this,

what? be specific

File: docs/guide/partitioning.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        num_buckets=4,
+        secondary_indices="B",
+    ).compute()
+    sorted(dm.partitions.keys())
+
+    dm = dm.load_index("B", store_factory())
+    # When bucketing by a value, kartothek guarantees that only one file within
+    # a given partition needs to be loaded when querying a value, as long as
+    # you also built a secondary index on that column
+    sorted(dm.indices["B"].eval_operator("==", 1))
```
Comment (lr4d) [inline docs/guide/partitioning.rst]: nitpick, found this to be more clear

```suggestion
If you need more control over the size of files and the distribution within the files you can also ask for explicit bucketing of values.

When bucketing by a value, kartothek guarantees that only one file within
a given partition needs to be loaded when querying a value, as long as
you also built a secondary index on that column.

.. ipython:: python
    :okwarning:

    dm = update_dataset_from_ddf(
        ddf,
        dataset_uuid="with_bucketing",
        store=store_factory,
        partition_on="A",
        shuffle=True,
        table="table",
        bucket_by="B",
        num_buckets=4,
        secondary_indices="B",
    ).compute()
    sorted(dm.partitions.keys())

    dm = dm.load_index("B", store_factory())

    sorted(dm.indices["B"].eval_operator("==", 1))
```

File: docs/guide/mutating_datasets.rst (hunk 4 above)
Comment (fjetter) [reply docs/guide/mutating_datasets.rst]: Yes, works as expected

File: docs/guide/partitioning.rst (hunk 5 above)
Comment (fjetter) [reply docs/guide/partitioning.rst]: Rephrased and added more context. Please review again

**Comment by fjetter** [comment]
The content, especially for API, can and should be iterated over. I just wanted to do the first step

**Pull request merged**